lib/api_callable.js

```diff
@@ -377,6 +389,115 @@ Bundleable.prototype.call = function(apiCall, argument, settings) {
 };
 
 /**
+ * StreamProxy is a proxy to gRPC-streaming method.
+ *
+ * @private
+ * @constructor
+ * @param {StreamType} type - the type of gRPC stream.
+ */
+function StreamProxy(type) {
+  DuplexStream.call(this, {
+    objectMode: true,
+    readable: type !== gax.StreamType.CLIENT_STREAMING,
+    writable: type !== gax.StreamType.SERVER_STREAMING
+  });
+  this.type = type;
+  this._writeQueue = [];
+  this._isEndCalled = false;
+  this.on('finish', this._onFinish.bind(this));
+}
+
+util.inherits(StreamProxy, DuplexStream);
+
+StreamProxy.prototype._read = function(n) {
+  if (this.type === gax.StreamType.CLIENT_STREAMING) {
+    this.emit('error', new Error('stream is not readable'));
+  }
+};
+
+StreamProxy.prototype._write = function(chunk, encoding, callback) {
+  if (this.type === gax.StreamType.SERVER_STREAMING) {
+    this.emit('error', new Error('stream is not writable'));
+    return;
+  }
+  if (this.stream) {
+    this.stream.write(chunk);
+  } else {
+    this._writeQueue.push(chunk);
+  }
+  callback();
+};
+
+StreamProxy.prototype._onFinish = function() {
+  if (this.stream) {
+    console.log(this.stream);
+    this.stream.end();
+  } else {
+    this._isEndCalled = true;
+  }
+};
+
+/**
+ * Specifies the target stream.
+ * @param {Stream} stream - the target stream.
+ */
+StreamProxy.prototype.setStream = function(stream) {
+  this.stream = stream;
+  stream.on('error', this.emit.bind(this, 'error'));
+  if (this.type !== gax.StreamType.CLIENT_STREAMING) {
+    stream.on('data', this.emit.bind(this, 'data'));
+    // Pushing null causes an ending process of the readable stream.
+    stream.on('end', this.push.bind(this, null));
+  }
+  if (this.type !== gax.StreamType.SERVER_STREAMING) {
+    this._writeQueue.forEach(function(data) {
+      stream.write(data);
+    });
+    this._writeQueue = [];
+    if (this._isEndCalled) {
+      stream.end();
+    }
+  }
+};
+
+/**
+ * An API caller for methods of gRPC streaming.
+ * @private
+ * @constructor
+ * @param {StreamDescriptor} descriptor - the descriptor of the method structure.
+ */
+function GrpcStreamable(descriptor) {
+  this.descriptor = descriptor;
+}
+
+GrpcStreamable.prototype.init = function(settings, callback) {
+  this.callback = callback;
+  this.stream = new StreamProxy(this.descriptor.type);
+  return this.stream;
+};
+
+GrpcStreamable.prototype.wrap = function(func) {
+  if (this.descriptor.type === gax.StreamType.CLIENT_STREAMING) {
+    return function(argument, metadata, options, callback) {
+      return func(metadata, options, callback);
+    };
+  } else if (this.descriptor.type === gax.StreamType.BIDI_STREAMING) {
+    return function(argument, metadata, options, callback) {
+      return func(metadata, options);
+    };
+  }
+  return func;
+};
+
+GrpcStreamable.prototype.call = function(apiCall, argument, settings) {
+  this.stream.setStream(apiCall(argument, this.callback));
```

Comment:
Won't this override any previous calls to the method?


```js
// ex:
var streamA = api.MyStreamingMethod({message: uniqueReqA});
var streamB = api.MyStreamingMethod({message: uniqueReqB});
```

`streamA` will be referencing the same stream as `streamB` which would cause problems if the user was expecting different streams based on the request message.